source/coreml.js

```diff
@@ -5,20 +5,34 @@ var protobuf = protobuf || require('./protobuf');
 
 coreml.ModelFactory = class {
 
-    match(context) {
-        const tags = context.tags('pb');
-        if (tags.get(1) === 0 && tags.get(2) === 2) {
-            return true;
+    _extractMLModelFromPackage(context) {
+        const mlmodels = context.entries('zip')
+            .filter(entry => !entry.name.startsWith('__MACOSX'))
+            .filter(entry => entry.name.endsWith('.mlmodel'));
+        if (mlmodels.length > 0) {
+            return mlmodels[0];
+        } else {
+            return null;
+        }
+    }
+
+    _isMLPackage(context) {
+        // match .mlpackage format -- the folder-like structure is interpreted as .mlpackage.zip
```

Comment:
I'm actually not sure; it seemed to happen automatically when I tested locally. I'm guessing that because it's a bundle (folder displayed as though it's a single file) on macOS, the File Open dialog allows you to select it (when whitelisting its file extension), and the webpage seems to receive the resulting selected file as a single .zip stream instead of a folder. I'm not sure if this behavior comes from macOS itself, Chromium, or a combination of the two.

Reply:
Dragging in Chrome as well as as using the Electron app this does not work. This will likely require implementing folder support as part of the pull requests and [Drag-and-drop-a-folder-onto-Chrome-now-available](https://developers.google.com/web/updates/2012/07/Drag-and-drop-a-folder-onto-Chrome-now-available). Has there been some discussion of the pros vs. cons of introducing `.mlpackage` bundles?

Reply:
> Drag & drop in Chrome as well as as using the Electron app does not work. It will likely require implementing folder support as part of the pull requests and Drag-and-drop-a-folder-onto-Chrome-now-available.

I see -- thanks for the pointer. Sorry I missed those cases; I'll look into implementing that for this PR.

> Has there been some discussion of the pros and cons of introducing .mlpackage bundles?

Do you mean introducing bundle support for Netron, or introducing mlpackage bundles at all? If the former, no, I was simply assuming we should add support for it. If the latter, while I'm sure there was internal discussion, I don't think any discussion has happened publicly yet, but if you have feedback on the use of a bundle I would encourage you to open an issue in [coremltools](https://github.com/apple/coremltools) and discuss there.

Reply:
Let me rework this a bit and split the pull request into multiple changes.

1) `coreml.ModelFactory.match` should change to operate at the file level and detect `Manifest.json` in a bundle.
2) The ML program const folding should follow the same approach as optimizing the TensorFlow Event File for PyTorch.
3) There will be additional changes for supporting weights files.

Once that is done the outer layer which is looking at every load context first as content of an archive could be adapted to support folders. That requires changes in a lot of places so not sure how soon that will happen.

Reply:
Thanks @lutzroeder! I appreciate all the effort. Please let me know if I can help by refactoring or splitting functionality from the commit in this PR.